mongo/integration/sdam_prose_test.go

```diff
@@ -232,4 +233,56 @@ func TestServerHeartbeatStartedEvent(t *testing.T) {
 		}
 		assert.Equal(t, expectedEvents, actualEvents)
 	})
+
+	mt := mtest.New(t)
+
+	mt.Run("polling must await frequency", func(mt *mtest.T) {
+		// Create a client with  heartbeatFrequency=100ms,
+		// serverMonitoringMode=poll. Use SDAM to record the number of times the
+		// a heartbeat is started.
+		var heartbeatStartedCount atomic.Int64
+
+		serverMonitor := &event.ServerMonitor{
+			ServerHeartbeatStarted: func(*event.ServerHeartbeatStartedEvent) {
+				heartbeatStartedCount.Add(1)
+			},
+		}
+
+		// Set the heartbeatInterval to a value so that 1/4 * value is long enough
+		// to ensure that the driver makes it's initial handshake with all servers
+		// in the deployment.
+		const heartbeatInterval = 200 * time.Millisecond
+
+		mt.ResetClient(options.Client().
+			SetServerMonitor(serverMonitor).
+			SetHeartbeatInterval(heartbeatInterval).
+			SetServerMonitoringMode(options.ServerMonitoringModePoll))
+
+		// Check on the count 4 times.
+		ticker := time.NewTicker(heartbeatInterval / 4)
+		t.Cleanup(ticker.Stop)
+
+		timer := time.NewTimer(heartbeatInterval - 1)
```

Comment:
Optional: This seems to rely on precise sleep synchronization and may intermittently fail if the scheduler responds to the `timer` goroutine slightly after the `ticker` goroutine (on the 4th "tick"). Consider using a more reliable assertion, like asserting a maximum number of heartbeats during an elapsed time.

E.g.
```go
start := time.Now()
time.Sleep(heartbeatInterval * 4)
got := heartbeatStartedCount.Load()

elapsed := time.Since(start)
wantMax := elapsed / heartbeatInterval

assert.LessOrEqual(mt,
	got,
	wantMax,
	"expected fewer than %d heartbeats in %v",
	wantMax,
	elapsed)
```
It may also be helpful to further reduce the heartbeat interval to speed up the test.

Reply:
Good point! A more robust solution would be to keep track of the discovered servers and assert that the number of handshakes align with that value within the `minHeartbeatFrequencyMS`.

Reply:
Ah, right, you'd have to know how many servers connected 🤔 . I believe you could derive that from the heartbeat events by counting distinct addresses.